src/v/cloud_storage/remote_partition.cc

```diff
@@ -336,12 +360,42 @@ class partition_record_batch_reader_impl final
             // base_offset and proceed with the next one. The client will see a
             // gap. The caching of the _segments.end() prevents this.
             if (_it == _end) {
+                // This branch might be taken if the last segment was merged
+                // with the previous one. (TODO: fix this)
```

Comment:
This change is a first step to enable SI for compacted topics.  There is a new PR in the works which will remove `_segments` collection entirely. The record batch reader will be using different API. Instead of storing an iterator it will be requesting `remote_partition` which will be returning proper reader for it (depending on an offset range). So the top level reader will only have to store an offset range that it's interested in.

Reply:
> So the top level reader will only have to store an offset range that it's interested in.

Yeah that's what I was suggesting. Cool stuff.